src/main/java/duke/data/Storage.java

```diff
@@ -0,0 +1,110 @@
+    public Storage(String fileName) {
+        this.dataFile = new File(fileName);
+    }
+
+    /**
+     * Saves the list to the dataFile stored locally.
+     * Parses the tasks and separates the different parameters with " | " for ease of parsing
+     */
+    public void saveToFile() {
+        try {
+            FileWriter myWriter = new FileWriter(dataFile, false);
+
+            // Loop through the taskList and parse the tasks to write it in the dataFile
+            for (Task current : taskList) {
+                myWriter.write(current.getTaskType() + Ui.DELIMITER + current.getDoneStatus() + Ui.DELIMITER
+                        + current.getDescription() + Ui.DELIMITER + current.getDueDate() + "\n");
+            }
+            myWriter.close(); // Must close the writer after each use
+
+            System.out.println("Successfully wrote to the file.");
+        } catch (IOException e) {
+            System.out.println("An error occurred.");
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Retrieves the data from the dataFile and returns it as an ArrayList of Tasks.
+     *
+     * @return ArrayList representing the tasks in the dataFile
+     */
+    public ArrayList<Task> loadFromFile() {
+        try { // Make a new file if possible
+            dataFile.getParentFile().mkdirs();
+            dataFile.createNewFile();
+
+
+            // Read the txt into the taskList arrayList
+            Scanner taskListReader = new Scanner(dataFile);
+
+            // While there are more tasks in the file
+            while (taskListReader.hasNext()) {
+                String taskDetails = taskListReader.nextLine();
+
+                // [0] is the Task category, [1] is isDone boolean, [2] is task desc, [3] is task dueDate
+                int isDoneFlagIndex = 1;
+                int taskDescIndex = 2;
+                int dueDateTimeIndex = 3;
+
+                if (taskDetails.startsWith("D ")) {
+                    taskList.add(new Deadline(taskDetails.split(Ui.SPLIT_DELIMITER)[taskDescIndex],
+                            LocalDate.parse(taskDetails.split(Ui.SPLIT_DELIMITER)[dueDateTimeIndex].split(" ")[0]),
+                            LocalTime.parse(taskDetails.split(Ui.SPLIT_DELIMITER)[dueDateTimeIndex].split(" ")[1]),
+                            !taskDetails.split(Ui.SPLIT_DELIMITER)[isDoneFlagIndex].equals("0")));
+
+                } else if (taskDetails.startsWith("E ")) {
+                    taskList.add(new Event(taskDetails.split(Ui.SPLIT_DELIMITER)[taskDescIndex],
+                            taskDetails.split(Ui.SPLIT_DELIMITER)[dueDateTimeIndex],
+                            !taskDetails.split(Ui.SPLIT_DELIMITER)[isDoneFlagIndex].equals("0")));
+
+                } else if (taskDetails.startsWith("T ")) {
+                    taskList.add(new Todo(taskDetails.split(Ui.SPLIT_DELIMITER)[taskDescIndex],
+                            !taskDetails.split(Ui.SPLIT_DELIMITER)[isDoneFlagIndex].equals("0")));
```

Comment:
Although you separated the lines, this part of code still seems a little bit hard to understand. Maybe we can consider separate them more.